src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -2314,17 +2267,15 @@ else if (t != null) {       // check steal chain
                             }
                             if ((s = task.status) < 0)
                                 break outer;            // validate
-                            if (q.source == sq && q.base == b && a[k] == t) {
-                                int nb = b + 1, nk = nb & (cap - 1);
+                            if (q.source == sq && q.base == b &&
+                                U.getReference(a, k) == t) {
                                 if (!eligible) {        // revisit if nonempty
-                                    if (!rescan && t == null &&
-                                        (a[nk] != null || q.top - b > 0))
+                                    if (!rescan && t == null && q.top - b > 0)
                                         rescan = true;
                                     break;
                                 }
-                                if (U.compareAndSetReference(
-                                        a, slotOffset(k), t, null)) {
-                                    q.updateBase(nb);
+                                if (U.compareAndSetReference(a, k, t, null)) {
+                                    q.base = b + 1;
                                     w.source = j;
```

Comment:
Might be worth commenting the volatile-write piggybacking which allows to get away with not calling updateBase above? 🤔